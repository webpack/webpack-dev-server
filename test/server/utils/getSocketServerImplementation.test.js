'use strict';

/* eslint-disable constructor-super, no-empty-function, no-useless-constructor, no-unused-vars, class-methods-use-this */

const getSocketServerImplementation = require('../../../lib/utils/getSocketServerImplementation');
const BaseServer = require('../../../lib/servers/BaseServer');
const SockJSServer = require('../../../lib/servers/SockJSServer');

describe('getSocketServerImplementation util', () => {
  it("should work with string serverMode ('sockjs')", () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: 'sockjs',
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should work with serverMode (SockJSServer class)', () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: SockJSServer,
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should work with serverMode (SockJSServer full path)', () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: require.resolve('../../../lib/servers/SockJSServer'),
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should work with serverMode (additional class methods)', () => {
    let result;

    const ExtendedSockJSServer = class ExtendedSockJSServer extends SockJSServer {
      myMethod() {
        this.test = true;
      }
    };

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: ExtendedSockJSServer,
      });
    }).not.toThrow();

    expect(result).toEqual(ExtendedSockJSServer);
  });

  const ClassWithoutConstructor = class ClassWithoutConstructor {};
  // eslint-disable-next-line no-undefined
  ClassWithoutConstructor.prototype.constructor = undefined;

  const badSetups = [
    {
      title: 'should throw with serverMode (bad path)',
      config: {
        serverMode: '/bad/path/to/implementation',
      },
    },
    {
      title:
        'should throw with serverMode (no constructor, send, close, onConnection methods)',
      config: {
        serverMode: ClassWithoutConstructor,
      },
    },
    {
      title:
        'should throw with serverMode (no send, close, onConnection methods)',
      config: {
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }
        },
      },
    },
    {
      title: 'should throw with serverMode (no close, onConnection methods)',
      config: {
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }

          send(connection, message) {}
        },
      },
    },
    {
      title: 'should throw with serverMode (no onConnection method)',
      config: {
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }

          send(connection, message) {}

          close(connection) {}
        },
      },
    },
  ];

  badSetups.forEach((setup) => {
    it(setup.title, () => {
      let thrown = false;
      try {
        getSocketServerImplementation(setup.config);
      } catch (e) {
        thrown = true;
        expect(e).toMatchSnapshot();
      }
    });
  });
});
