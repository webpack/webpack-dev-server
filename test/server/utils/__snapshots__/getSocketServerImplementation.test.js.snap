// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSocketServerImplementation util should throw with serverMode (bad path) 1`] = `[Error: serverMode must be a string denoting a default implementation (e.g. 'sockjs'), a full path to a JS file which exports a class extending BaseServer (webpack-dev-server/lib/servers/BaseServer) via require.resolve(...), or the class itself which extends BaseServer]`;

exports[`getSocketServerImplementation util should throw with serverMode (no close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options - serverMode must have a close(connection) method that closes the provided client connection object
options - serverMode must have a onConnection(f) method that calls f(connection) whenever a new client connection is made
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no constructor, send, close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options [object Object]
options - serverMode must have a send(connection, message) method that sends the message string to the provided client connection object
options - serverMode must have a close(connection) method that closes the provided client connection object
options - serverMode must have a onConnection(f) method that calls f(connection) whenever a new client connection is made
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no onConnection method) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options - serverMode must have a onConnection(f) method that calls f(connection) whenever a new client connection is made
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no send, close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options - serverMode must have a send(connection, message) method that sends the message string to the provided client connection object
options - serverMode must have a close(connection) method that closes the provided client connection object
options - serverMode must have a onConnection(f) method that calls f(connection) whenever a new client connection is made
]
`;
